src/finalizer/finalizer.cpp

```diff
@@ -5,207 +5,325 @@
+                            *pbRestartRequired = TRUE;
+                        }
+
+                        // Reset potential failures so we can continue to remove as many dependents as possible.
+                        hr = S_OK;
+                    }
+                    else if (HRESULT_FROM_WIN32(ERROR_UNKNOWN_PRODUCT) == hr || HRESULT_FROM_WIN32(ERROR_UNKNOWN_PROPERTY) == hr)
+                    {
                         // Possibly a corrupted provider key that wasn't cleaned up. We'll just ignore it.
-						LogStringLine(REPORT_STANDARD, "  Product is not installed, ProductCode:%ls, result: 0x%.8x", sczProductId, hr);
-						hr = S_OK;
-					}
-				}
-			}
-		}
+                        LogStringLine(REPORT_STANDARD, "  Product is not installed, ProductCode:%ls, result: 0x%.8x", sczProductId, hr);
+                        hr = S_OK;
+                    }
+                }
+            }
+        }
 
-		ReleaseRegKey(hkDependentsKey);
-		ReleaseRegKey(hkProviderKey);
-	}
+        ReleaseRegKey(hkDependentsKey);
+        ReleaseRegKey(hkProviderKey);
+    }
 
 LExit:
-	ReleaseStr(sczProductName);
-	ReleaseStr(sczProductId);
-	ReleaseStr(sczProviderKey);
-	ReleaseStr(sczDependentsKey);
-	ReleaseRegKey(hkDependentsKey);
-	ReleaseRegKey(hkProviderKey);
-	ReleaseRegKey(hkInstallerDependenciesKey);
-	return hr;
+    ReleaseStr(sczProductName);
+    ReleaseStr(sczProductId);
+    ReleaseStr(sczProviderKey);
+    ReleaseStr(sczDependentsKey);
+    ReleaseRegKey(hkDependentsKey);
+    ReleaseRegKey(hkProviderKey);
+    ReleaseRegKey(hkInstallerDependenciesKey);
+    return hr;
+}
+
+extern "C" HRESULT ParseSdkVersion(LPWSTR sczSdkVersion, INT * piMajor, INT * piMinor, INT * piFeatureBand)
+{
+    HRESULT hr = S_OK;
+    UINT cVersionParts = 0;
+    LPWSTR* rgsczVersionParts = NULL;
+
+    hr = StrSplitAllocArray(&rgsczVersionParts, &cVersionParts, sczSdkVersion, L".");
+    ExitOnFailure(hr, "Failed to split version.");
+
+    // We only care about the major.minor.patch values
+    // to convert to a feature band. If we don't have at least
+    // all 3 parts, we'll ignore the value.
+    if (3 > cVersionParts)
+    {
+        ExitOnFailure(E_INVALIDARG, "Invalid SDK version: %ls %li", sczSdkVersion, cVersionParts);
+    }
+
+    int iMajor = 0;
+    int iMinor = 0;
+    int iPatch = 0;
+
+    hr = StrStringToInt32(rgsczVersionParts[0], 0, piMajor);
+    ExitOnFailure(hr, "Invalid major version.");
+    hr = StrStringToInt32(rgsczVersionParts[1], 0, piMinor);
+    ExitOnFailure(hr, "Invalid minor version.");
+
+    // If this is a valid SDK version the 'patch' should be a 3 digit field
+    // containing the feature band and patch level, e.g. 100 or 207. We 
+    // can discard any prerelease labels from the semantic version.
+    hr = StrStringToInt32(rgsczVersionParts[2], 3, &iPatch);
+    ExitOnFailure(hr, "Invalid patch version.");
+
+    if (100 > iPatch)
```

Comment:
I think this is wrong.  If the version number is 6.0.205, then `iPatch` here would be 205.  If you want to know the actual "SDK patch" number, then you'd subtract `*piFeatureBand` from `iPatch`.  But we don't need that value here.

Reply:
I don't care about the actual patch, I just want the feature band portion.

Here's an excerpt from the finalizer log when I tested locally to clean up 6.0.300 (which I created fake installers and registry keys for). The second entry is probably the closest to your example. All I want to do is determine whether the band the finalizer is cleaning up is actually installed, so in this case, it maps 5.0.104 to 5.0.100 and that gets compared against 6.0.300 and subsequently skipped.

```
0BB8:2E24][2021-06-28T08:27:22]i000: Detecting installed SDK versions for 6.0.300
[0BB8:2E24][2021-06-28T08:27:22]i000: Scanning SOFTWARE\WOW6432Node\dotnet\Setup\InstalledVersions\x64\sdk
[0BB8:2E24][2021-06-28T08:27:22]i000: SDK version detected: 6.0.100-preview.2.21155.3, mapping to 6.0.100.
[0BB8:2E24][2021-06-28T08:27:22]i000: SDK version detected: 5.0.104, mapping to 5.0.100.
[0BB8:2E24][2021-06-28T08:27:22]i000: SDK version detected: 5.0.300-preview.21180.15, mapping to 5.0.300.
[0BB8:2E24][2021-06-28T08:27:22]i000: SDK version detected: 2.1.816, mapping to 2.1.800.
[0BB8:2E24][2021-06-28T08:27:22]i000: SDK version detected: 2.1.524, mapping to 2.1.500.
[0BB8:2E24][2021-06-28T08:27:22]i000: SDK version detected: 3.1.410, mapping to 3.1.400.
[0BB8:2E24][2021-06-28T08:27:22]i000: SDK with feature band 6.0.300 could not be found.
```

Reply:
Nevermind, I would have written the condition as `iPatch < 100`, so I got confused about the comparison direction.